### PR TITLE
feat(Store): Implement Store

### PR DESCRIPTION
**Pull request by Khaazz**
# PULL REQUEST

## **Overview**

This PR implements `Store` DataStructure.
This structure brings all utility function to manipulate `Key => Value` structures in one place.
This also correctly separate data manipulation function from the actual cache holder. It now appears under the `cache` property in Store.
A Store is iterable, and you can apply `get`, `set`, `delete`, `has` (along with all others utility) functions directly on it.

Store is used in all registries directly. By extension, registries are iterable and you can use all Store methods directly on it.
Collection is just a Store that use a Map and enforce a particular type of object to be used.

**Why Store, Why not directly extending Map?**
- Store can be used with any Key-Value store that implements `has`, `get`, `set`, `delete`, and `keys`, `values`, `entries`.
- Store allow to entirely separate the utility function from the actual storage
- Store prevent from overriding `Map` methods.
- Extending Store (eg: for registries) is fine since we can easily override specific methods. We also keep separation between the Structure that handles the data and the structure that holds the data.

## **Status**

- [X] Typings have been updated or don't need to be.
- [X] This PR have been tested and is ready to be merged.

## **Semantic versioning classification**

- [ ] This PR introduces *BREAKING* changes.
- [X] This PR adds new features, improve the code and implies minimal changes.
- [ ] This PR fixes a bug and references the relevant issue or documentation.
- [X] This PR improve performance or code refactor without API changes.
- [ ] This PR **only** includes non-code changes (documentation, CI, tools...).

### CODE REVIEW

File: src/Utility/Store.js [hunk 1]
```diff
@@ -0,0 +1,311 @@
+/**
+ * Generic class that expose all data manipulation methods.
+ * A Store can be constructed with any Key => Value structure (Map like object).
+ * It exposes all methods similar to aarray methods in one place.
+ *
+ * @author KhaaZ
+ *
+ * @template T
+ * @class Store
+ * @prop {Map} cache - Key => Value Data structure
```
Comment (bsian03) [inline src/Utility/Store.js]: Generics
```suggestion
 * @prop {Map<string, T>} cache - Key => Value Data structure
```

File: src/Utility/Store.js [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+     * @memberof Store
+     */
+    get size() {
+        return this.cache.size;
+    }
+
+    /**
+     * Iterator over cache.entries.
+     *
+     * @returns {Iterator<Key, Value>}
```
Comment (bsian03) [inline src/Utility/Store.js]:
```suggestion
     * @returns {IterableIterator<[string, T]>}
```

File: src/Utility/Store.js [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+     * @readonly
+     * @type {Number}
+     * @memberof Store
+     */
+    get size() {
+        return this.cache.size;
+    }
+
+    /**
+     * Iterator over cache.entries.
```
Comment (bsian03) [inline src/Utility/Store.js]:
```suggestion
     * Iterator over `cache.entries`
```
Just for JSDoc markdown

File: src/Utility/Store.js [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+     *
+     * @returns {Iterator<Key, Value>}
+     * @memberof Store
+     */
+    entries() {
+        return this.cache.entries();
+    }
+
+    /**
+     * Iterator over cache.keys
```
Comment (bsian03) [inline src/Utility/Store.js]:
```suggestion
     * Iterator over `cache.keys`
```

File: src/Utility/Store.js [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+     *
+     * @returns {Iterator<Key>}
```
Comment (bsian03) [inline src/Utility/Store.js]:
```suggestion
     * @returns {IterableIterator<T>}
```

File: src/Utility/Store.js [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+     * @memberof Store
+     */
+    keys() {
+        return this.cache.keys();
+    }
+
+    /**
+     * Iterator over cache.values
```
Comment (bsian03) [inline src/Utility/Store.js]:
```suggestion
     * Iterator over `cache.values`
```

File: src/Utility/Store.js [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+     *
+     * @returns {Iterator<Value>}
```
Comment (bsian03) [inline src/Utility/Store.js]:
```suggestion
     * @returns {IterableIterator<T>}
```

File: src/Utility/Store.js [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+     *
+     * @returns {Iterator}
+     * @memberof Store
+     */
+    [Symbol.iterator]() {
+        return this.cache[Symbol.iterator]();
+    }
+
+    /**
+     * Wether an element exists in the Store
```
Comment (bsian03) [inline src/Utility/Store.js]:
```suggestion
     * Whether an element exists in the Store
```

File: src/Utility/Store.js [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+    set(key, value) {
+        this.cache.set(key, value);
+        return this;
+    }
+
+    /**
+     * Delete an element by key in the Store
+     *
+     * @param {String} key
+     * @returns {Boolean} - Whether the element was suppressed or not
```
Comment (bsian03) [inline src/Utility/Store.js]: just use `deleted` lol

File: src/Utility/Store.js [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+     */
+    delete(key) {
+        return this.cache.delete(key);
+    }
+
+    // General cache manipulation
+
+    /**
+     * Store.cache to array
+     * [ value, value, value ]
```
Comment (bsian03) [inline src/Utility/Store.js]:
```suggestion
     * `[ value, value, value ]`
```
Though idk why this is needed

File: src/Utility/Store.js [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+     * @memberof Store
+     */
+    delete(key) {
+        return this.cache.delete(key);
+    }
+
+    // General cache manipulation
+
+    /**
+     * Store.cache to array
```
Comment (bsian03) [inline src/Utility/Store.js]:
```suggestion
     * `Store.cache` to array
```

File: src/Utility/Store.js [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+     */
+    toArray() {
+        return [...this.values()];
+    }
+
+    /**
+     * Store to object
+     * { key: value, key: value }
+     *
+     * @returns {{[key:string]: T}}
```
Comment (bsian03) [inline src/Utility/Store.js]:
```suggestion
     * @returns {Object.<string, T>}
```
Just so it's clearer to see

File: src/Utility/Store.js [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+     * @returns {Array<T>}
+     * @memberof Store
+     */
+    toArray() {
+        return [...this.values()];
+    }
+
+    /**
+     * Store to object
+     * { key: value, key: value }
```
Comment (bsian03) [inline src/Utility/Store.js]:
```suggestion
     * `{ key: value, key: value }`
```
As above

File: src/Utility/Store.js [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+     *
+     * @returns {Array<T>}
+     * @memberof Store
+     */
+    toArray() {
+        return [...this.values()];
+    }
+
+    /**
+     * Store to object
```
Comment (bsian03) [inline src/Utility/Store.js]:
```suggestion
     * `Store.cache` to object
```
Consistency

File: src/Utility/Store.js (hunk 10 above)
Comment (Khaazz) [reply src/Utility/Store.js]: Probably not needed, I'll just remove it

File: src/Utility/Store.js [hunk 15]
```diff
@@ -0,0 +1,311 @@
...
+        return obj;
+    }
+
+    /**
+     * Apply a function to the Store and returns a new Store.
+     * Usable over: map, filter, some, every, forEach, forEachAsync
+     *
+     * @param {String} key - The property to use as key for the new Store
+     * @param {String} func - The function name to apply to the Store
+     * @param {...any} args - All the argument that need to be applied to the function
```
Comment (bsian03) [inline src/Utility/Store.js]:
```suggestion
     * @param {...any} args - All the arguments that need to be applied to the function
```

File: src/Utility/Store.js (hunk 13 above)
Comment (Khaazz) [reply src/Utility/Store.js]: I'll remove that as well

File: src/Utility/Store.js [hunk 16]
```diff
@@ -0,0 +1,309 @@
...
+     */
+    apply(key, func, ...args) {
+        return new Store(new Map(this[func](...args).map(e => [e[key], e] ) ) );
+    }
+
+    /**
+     * Execute a function over each element of the Store in a synchronous fashion
+     *
+     * @param {(i: T) => void} func - A function that takes an object and do something on it
+     * @returns {Store} - returns the current Store
```
Comment (bsian03) [inline src/Utility/Store.js]:
```suggestion
     * @returns {Store} The current Store
```

File: src/Utility/Store.js [hunk 17]
```diff
@@ -0,0 +1,309 @@
...
+            func(item);
+        }
+        return this;
+    }
+
+    /**
+     * Execute a function over each element of the Store in an asynchronous fashion
+     *
+     * @param {(i: T) => Promise<void>} func - A function that takes an object and do something on it
+     * @returns {Promise<Store>} - returns the current Store
```
Comment (bsian03) [inline src/Utility/Store.js]:
```suggestion
     * @returns {Promise<Store>} - The current Store
```

File: src/Utility/Store.js [hunk 18]
```diff
@@ -0,0 +1,309 @@
...
+        for (const item of this.values() ) {
+            if (func(item) ) {
+                arr.push(item);
+            }
+        }
+        return arr;
+    }
+
+    /**
+	 * Reduce values by function
```
Comment (bsian03) [inline src/Utility/Store.js]:
```suggestion
     * Reduce values by function
```

File: src/Utility/Store.js [hunk 19]
```diff
@@ -0,0 +1,309 @@
...
+     *
+     * @template U
+     * @param {(accumulator: U, val: T) => U} func - Function to execute on each element in the array
+     * @param {Number} [initialValue=0] - Value to use as the first argument to the first call of the callback
+     * @returns {U} Accumulator
+     * @memberof Store
+	 */
```
Comment (bsian03) [inline src/Utility/Store.js]:
```suggestion
     */
```

File: src/Utility/Store.js [hunk 20]
```diff
@@ -0,0 +1,309 @@
...
+            result = func(result, val);
+        }
+        return result;
+    }
+
+    /**
+     * Test if at least one element passes the test implemented by the provided function. Returns true if yes, or false if not.
+     *
+     * @param {(i: T) => Boolean} func - A function that takes an object and returns true if it matches
+     * @returns {Boolean} An array containing all the objects that matched
```
Comment (bsian03) [inline src/Utility/Store.js]:
```suggestion
     * @returns {Boolean} An array containing all the objects that matched
```
```suggestion
     * @returns {Boolean} Whether or not any of the elements returned true
```

File: src/Utility/Store.js [hunk 21]
```diff
@@ -0,0 +1,309 @@
...
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Test if all elements pass the test implemented by the provided function. Returns true if yes, or false if not.
+     *
+     * @param {(i: T) => Boolean} func - A function that takes an object and returns true if it matches
+     * @returns {Boolean} An array containing all the objects that matched
```
Comment (bsian03) [inline src/Utility/Store.js]:
```suggestion
     * @returns {Boolean} Whether or not all of the elements returned true
```

File: src/Utility/Store.js [hunk 22]
```diff
@@ -0,0 +1,309 @@
...
+     * Get a random object from the Store
+     *
+     * @returns {T?} The random object, or null if there is no match
+     * @memberof Store
+     */
+    random() {
+        if (!this.size) {
+            return null;
+        }
+        return Array.from(this.values() )[Math.floor(Math.random() * this.size)];
```
Comment (bsian03) [inline src/Utility/Store.js]: Why not just
```js
return this.toArray()[Math.floor(Math.random() * this.size)];
```

File: src/Utility/Store.js [hunk 23]
```diff
@@ -0,0 +1,309 @@
...
+        const obj = {};
+        for (const [key, value] of this.entries() ) {
+            obj[key] = value;
+        }
+        return obj;
+    }
+
+    /**
+     * Apply a function to the Store and returns a new Store.
+     * Usable over: map, filter, some, every, forEach, forEachAsync
```
Comment (bsian03) [inline src/Utility/Store.js]:
```suggestion
     * Usable over: `map`, `filter`, `toArray`
```

File: src/Utility/Store.js (hunk 22 above)
Comment (Khaazz) [reply src/Utility/Store.js]: Very true

File: src/Utility/Collection.js [hunk 24]
```diff
@@ -1,16 +1,17 @@
+import Store from './Store';
+
 /**
- * Extended Map with built in methods for ease of data manipulation.
- * Based on Eris.Collection
+ * Custom Store that uses Map as cache.
+ * Enforce that a single object type exists in a Collection.
```
Comment (bsian03) [inline src/Utility/Collection.js]: this doesn't make any sense

Comment (bsian03) [reply src/Utility/Collection.js]:
```suggestion
 * Enforces that a single object type exists in a Collection.
```

File: src/Utility/External/LRUCache.js [hunk 25]
```diff
@@ -30,30 +30,27 @@ class Node {
  *
  * @author KhaaZ
  *
- * @class LRUCache
- *
  * @template T
- *
+ * @class LRUCache
+ * @extends Store<Node>
```
Comment (bsian03) [inline src/Utility/External/LRUCache.js]:
```suggestion
 * @extends Store<Node<T>>
```

File: src/Utility/External/LRUCache.js [hunk 26]
```diff
@@ -30,30 +30,27 @@ class Node {
...
  */
-class LRUCache {
+class LRUCache extends Store {
     /**
      * Creates an instance of LRUCache.
      *
-     * @param {Number} limit - Max number of element in the Collection
-     * @param {Object} options - Options used to construct the Collection
-     * @param {new (...args: Array<any> ) => T} [options.base=null]
+     * @param {Number} limit - Max number of element in the cache
```
Comment (bsian03) [inline src/Utility/External/LRUCache.js]:
```suggestion
     * @param {Number} limit - Max number of elements in the cache
```

File: src/Utility/External/LRUCache.js [hunk 27]
```diff
@@ -143,8 +134,7 @@ class LRUCache {
             this.tail = node.prev;
         }
   
-        this._cache.delete(key);
-        this.size--;
+        this.cache.delete(key);
```
Comment (bsian03) [inline src/Utility/External/LRUCache.js]:
```suggestion
        return this.cache.delete(key);
```
Store returns a boolean for this

File: src/Utility/Store.js [hunk 28]
```diff
@@ -0,0 +1,309 @@
...
+     * @memberof Store
+     */
+    values() {
+        return this.cache.values();
+    }
+
+    /**
+     * Iterator over cache
+     *
+     * @returns {Iterator}
```
Comment (bsian03) [inline src/Utility/Store.js]:
```suggestion
     * @returns {IterableIterator<[string, T]>;}
```

File: src/Utility/External/LRUCache.js [hunk 29]
```diff
@@ -123,13 +94,33 @@ class LRUCache {
...
-     * Remove an element from the LRUCache
+     * Add a value in the LRU cache.
      *
      * @param {String} key
+     * @param {T} value
      * @memberof LRUCache
      */
-    remove(key) {
-        const node = this._cache.get(key);
+    set(key, value) {
```
Comment (bsian03) [inline src/Utility/External/LRUCache.js]: Store returns `Store<T>` for this

Comment (Khaazz) [reply src/Utility/External/LRUCache.js]: I'll have a look

**Pull request merged**